extensions/amp-story-auto-ads/0.1/story-ad-page.js

```diff
@@ -0,0 +1,486 @@
+    // TODO(ccordry): Move button to shadow root.
+    const a = this.doc_.createElement('a');
+    a.className = 'i-amphtml-story-ad-link';
+    a.setAttribute('target', '_blank');
+    setStyles(a, {
+      'font-size': '0',
+      opactiy: '0',
+      transform: 'scale(0)',
+    });
+    a.href = this.ctaUrl_;
+    a.textContent = this.ctaText_;
+
+    if (a.protocol !== 'https:' && a.protocol !== 'http:') {
+      user().warn(TAG, 'CTA url is not valid. Ad was discarded');
+      return false;
+    }
+
+    // Click listener so that we can fire `story-ad-click` analytics trigger at
+    // the appropriate time.
+    a.addEventListener('click', () => {
+      const vars = {
+        [AnalyticsVars.AD_CLICKED]: Date.now(),
+      };
+      this.analyticsEvent_(AnalyticsEvents.AD_CLICKED, vars);
+    });
+
+    const ctaLayer = this.doc_.createElement('amp-story-cta-layer');
+    ctaLayer.appendChild(a);
+    this.pageElement_.appendChild(ctaLayer);
+    return true;
+  }
+
+  /**
+   * Find all `amp4ads-vars-` prefixed meta tags and store them in single obj.
+   * @private
+   */
+  extractA4AVars_() {
+    const tags = getA4AMetaTags(this.adDoc_);
+    iterateCursor(tags, tag => {
+      const name = tag.name.split('amp4ads-vars-')[1];
+      const {content} = tag;
+      this.a4aVars_[name] = content;
+    });
+  }
+
+  /**
+   * TODO(#24080) Remove this when story ads have full ad network support.
+   * This in intended to be a temporary hack so we can can support
+   * ad serving pipelines that are reliant on using amp-ad-exit for
+   * outlinks.
+   * Reads amp-ad-exit config and tries to extract a suitable outlink.
+   * If there are multiple exits present, behavior is unpredictable due to
+   * JSON parse.
+   * @private
+   */
+  readAmpAdExit_() {
+    const ampAdExit = elementByTag(
+      dev().assertElement(this.adDoc_.body),
+      'amp-ad-exit'
+    );
+    if (ampAdExit) {
+      try {
+        const {children} = ampAdExit;
+        userAssert(
+          children.length == 1,
+          'The tag should contain exactly one <script> child.'
+        );
+        const child = children[0];
+        userAssert(
+          isJsonScriptTag(child),
+          'The amp-ad-exit config should ' +
+            'be inside a <script> tag with type="application/json"'
+        );
+        const config = assertConfig(parseJson(child.textContent));
+        const target = config['targets'][Object.keys(config['targets'])[0]];
+        this.ampAdExitOutlink_ = target['finalUrl'];
+      } catch (e) {
+        dev().error(TAG, e);
+      }
+    }
+  }
+
+  /**
+   * Create attribution if creative contains the appropriate meta tags.
+   * @private
+   */
+  maybeCreateAttribution_() {
+    const href = this.a4aVars_[A4AVarNames.ATTRIBUTION_URL];
+    const src = this.a4aVars_[A4AVarNames.ATTRIBUTION_ICON];
+
+    // Ad attribution is optional, but need both to render.
+    if (!href && !src) {
+      return;
+    }
+
+    if (!href || !src) {
```

Comment:
seems this check is redundant as we check both URLs in L436 anyway.

Reply:
This would change from warning to throwing when only 1 is present. Do you think error is better?

Reply:
attribution icon is necessary right? So I think we should throw.
but we need to make sure the exception is caught somewhere not causing the whole story to break. we should just skip that invalid ad

Reply:
Sounds good. Wrapped this method in a try/catch. I think this is the right thing to do, but it is a small behavior change. Previously we were showing the ad without the icon, now it will move on to the next ad.